cached_addr_book.go

```diff
@@ -0,0 +1,275 @@
+package main
+
+import (
+	"context"
+	"io"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/libp2p/go-libp2p/core/event"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/host/peerstore/pstoremem"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	probeDurationHistogram = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:      "probe_duration_seconds",
+		Namespace: "someguy",
+		Subsystem: "cached_addr_book",
+		Help:      "Duration of peer probing operations in seconds",
+		// Buckets optimized for expected probe durations from ms to full timeout
+		Buckets: []float64{0.5, 1, 2, 5, 10, 30, 60, 120},
+	})
+
+	peerStateSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name:      "peer_state_size",
+		Subsystem: "cached_addr_book",
+		Namespace: "someguy",
+		Help:      "Number of peers object currently in the peer state",
+	})
+)
+
+const (
+	// The TTL to keep recently connected peers for. Same as DefaultProviderAddrTTL in go-libp2p-kad-dht
+	RecentlyConnectedAddrTTL = time.Hour * 24
+
+	// Connected peers don't expire until they disconnect
+	ConnectedAddrTTL = math.MaxInt64
+
+	// How long to wait since last connection before probing a peer again
+	PeerProbeThreshold = time.Hour
```

Comment:
If you search for references to `peerstore.AddressTTL` you'll see it isn't used by anything (both in go-libp2p and go-libp2p-kad-dht) so maybe we should remove it?

There are two relevant address TTLs:
- [RecentlyConnectedTTL](https://github.com/libp2p/go-libp2p/blob/8423de3a64f17f6bec18bf57b472e5a3615883db/core/peerstore/peerstore.go#L29-L31) which is 15 minutes 
- [ProviderAddrTTL](https://github.com/libp2p/go-libp2p-kad-dht/blob/25e4aff78feaf7a83c0ebde80b538521210292b6/amino/defaults.go/#L49) which is 24 hours and is only for addresses associated with a provider record you are storing (so much less prevalent)


If we want to probe after the go-libp2p TTL expires, that would have to be 15 minutes for most addresses, but probing that frequently would mean we probe almost every peer every `probePeers` run, since that's how often the probe runs. Once we histogram data from production on how long `probePeers` takes, we can adjust.